### PR TITLE
documentation: adding type hints and doc-strings (1/n)

### PR DESCRIPTION
**Pull request by lawliet19189**
- Almost all of the changes are just to improve the readability of the code. 
- few exception handling changes
- few refactors to help linters identify potential bugs in development

### CODE REVIEW

File: dsp/modules/gpt3.py [hunk 1]
```diff
@@ -110,13 +152,13 @@ def __call__(self, prompt, only_completed=True, return_sorted=False, **kwargs):
 
 
 @CacheMemory.cache
-def cached_gpt3_request_v2(**kwargs):
-    return openai.Completion.create(**kwargs)
+def cached_gpt3_request_v2(**kwargs) -> OpenAIObject:
+    return cast(OpenAIObject, openai.Completion.create(**kwargs))
```
Comment (okhat) [inline dsp/modules/gpt3.py:156]: Thanks so much for doing this, Sri! For this function, any changes will unfortunately invalidate all old cached examples in CacheMemory due to way this library works. Ideally we'll migrate to a better cache in the future, but for now we don't want to do that.

File: dsp/modules/gpt3.py [hunk 2]
```diff
@@ -110,13 +152,13 @@ def __call__(self, prompt, only_completed=True, return_sorted=False, **kwargs):
...
 
 
 @functools.lru_cache(maxsize=None if cache_turn_on else 0)
 @NotebookCacheMemory.cache
-def cached_gpt3_request_v2_wrapped(**kwargs):
+def cached_gpt3_request_v2_wrapped(**kwargs) -> OpenAIObject:
```
Comment (okhat) [inline dsp/modules/gpt3.py:161]: same as a bove

File: dsp/modules/gpt3.py (hunk 1 above)
Comment (lawliet19189) [reply dsp/modules/gpt3.py:156]: Got it. My plan was to run the notebooks again, so caches would be updated. These aren't major changes, so dropping them is fine as well.

File: dsp/templates/template_v2.py [hunk 3]
```diff
@@ -1,45 +1,32 @@
 from collections import namedtuple
 import re
-
+from typing import Union, Any
 import dsp
+from dsp.primitives.demonstrate import Example
+from .utils import passages2text, format_answers
```
Comment (lawliet19189) [inline dsp/templates/template_v2.py:6]: review notes: removed redundant function definitions of passages2text and format_answers

File: dsp/primitives/demonstrate.py [hunk 4]
```diff
@@ -74,85 +79,78 @@ def do_augment(train, k=None, return_all=False):
...
-        and dsp.settings.branch_idx
-    )
-    branch_idx = branch_idx or 0
-    # print(f"branch_idx = {branch_idx}")
-
-    rng = random.Random(branch_idx)
-    shuffled_train = list(train)
+def sample(train: list[Example], k: int):
+    """Sample k examples from train."""
+    rng = random.Random(dsp.settings.branch_idx)
```
Comment (lawliet19189) [inline dsp/primitives/demonstrate.py:84]: reviewer note: branch_idx default value is set to 0 in `dsp.settings.brach_idx` during init, therefore the previous checks aren't necessary.

File: dsp/primitives/demonstrate.py [hunk 5]
```diff
@@ -74,85 +79,78 @@ def do_augment(train, k=None, return_all=False):
...
+    shuffled_train = [example.copy() for example in train]
     rng.shuffle(shuffled_train)
 
-    subset = shuffled_train[:k]
-    subset = [dsp.Example(x) for x in subset]  # augmented=False
```
Comment (lawliet19189) [inline dsp/primitives/demonstrate.py:91]: reviewer note: shuffle_train does not affect the input type, so Example() conversion is unnecessary.

**Comment by lawliet19189** [comment]
closing this in-favour of #28

**Pull request closed without merging**